compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -11623,131 +11623,145 @@ OMR::Z::TreeEvaluator::arraycmpEvaluator(TR::Node * node, TR::CodeGenerator * cg
    TR::Register * firstBaseReg = NULL;
    TR::Register * secondBaseReg = NULL;
    bool lenMinusOne=false;
-   TR_ASSERT(!node->isArrayCmpLen() || !node->isArrayCmpSign(), "Invalid arraycmp node");
-   if (!node->isArrayCmpLen())
+
+   // use CLC
+   TR::Register * resultReg;
+
+   if (elemsExpr->getOpCode().isLoadConst())
       {
-      // use CLC
-      TR::Register * resultReg;
+      int32_t elems = (int32_t) getIntegralValue(elemsExpr); //get number of elements (in bytes)
```

Comment:
Can you use `static_cast` here? Also just confirming, is `elemExps` `iconst` or `lconst`?

Reply:
`elemsExpr` is the length child of the `arraycmp`. An `arraycmp` is generated in one place in OMR [1] and one place in OpenJ9 [2]. In OMR it generates an `imul` for the length, and that would be optimized to `iconst` to hit this codegen path. In OpenJ9 the length child is wrapped in an `i2l` if the platform is 64-bit, so it is optimized to `lconst` on 64-bit platform but `iconst` on 31-bit to hit this path.

In conclusion:
64-bit: `elemsExpr` could be `iconst` or `lconst` depending on where the `arraycmp` is generated
31-bit: `elemsExpr` is `iconst`

I could use `int64_t` instead of `int32_t` for `elems` here since `getIntegralValue` returns `int64_t` and the `MemCmpConstLenSignMacroOp` constructor (where `elems` is used later) takes an `int64_t`.

[1] https://github.com/Spencer-Comin/omr/blob/26a4bc82a7fb76a3aaed44dc923086b59957f4c0/compiler/optimizer/LoopReducer.cpp#L2162
[2] https://github.com/Spencer-Comin/openj9/blob/c68f163bea76f3385d6d786d503b1c2dc4756af0/runtime/compiler/optimizer/IdiomTransformations.cpp#L9603